### ISSUE TITLE
Accuracy reduced when predicting from model trained by ML5

### ISSUE THREAD

**Issue by plat2eau** [tensorflow/tfjs#1940]
tfjs 1.0.0

Google Chrome | 76.0.3809.132 (Official Build) (64-bit) (cohort: 76_132_win)

I was using ML5 to train models for image classification in my project. I used the Feature Extractor for transfer learning. I was using mobilenet_v1_0.25 as a base model. I wanted to integrate it such that it performs predictions from chrome extension. I had to use tfjs because I found that ML5 does not run from the background page of the extension. I used tfjs to load models trained by ML5 and then start predicting. However, the prediction accuracy was very low in tfjs as compared to when predicting with the same model in ML5 itself. 

I am using `findMlMatches()` as the triggering function that takes an image as a parameter.

    // load tensorflow model and weights  
    async load() {
        console.log("ML Data loading..");
        // ! ==========================================
        // ! This is a work around and will only work for default version and alpha values that were used while training model.
        this.mobilenet = await tf.loadLayersModel("https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json");
        const layer = this.mobilenet.getLayer('conv_pw_13_relu');
        // ! ==========================================

        this.mobilenetFeatures = await tf.model({ inputs: this.mobilenet.inputs, outputs: layer.output });
        this.customModel = await tf.loadLayersModel("./model.json");
        this.model.add(this.mobilenetFeatures);
        this.model.add(this.customModel);
    }

    dispose() {
        if (this.model) {
            this.model.dispose();
        }
    }

    async findMlMatches(res, screenshot) {
        console.time('ML Match Time');
        return new Promise((resolve, reject) => {
            if (res != null) {
                resolve(res);
            }
            this.loadImage(screenshot)
                .then(image => {
                    let result = this.getTopKClasses(this.predict(image), 5)
                    console.log(result);
                    console.timeEnd('ML Match Time');
                    resolve(result)
                })
        });
    }

    loadImage(imageUrl) {
        return new Promise((resolve, reject) => {
            let img = new Image();
            img.crossOrigin = '';
            img.onload = () => {
                resolve(img);
            };
            img.onerror = () => {
                reject("Error: Unable to load image");
            };
            img.src = imageUrl;
        });
    }

    getTopKClasses(logits, topK) {
        const predictions = logits;
        const values = predictions.dataSync();
        predictions.dispose();
        let predictionList = [];
        for (let i = 0; i < values.length; i++) {
            predictionList.push({ value: values[i], index: i });
        }
        predictionList = predictionList
            .sort((a, b) => {
                return b.value - a.value;
            })
            .slice(0, topK);
        console.log(predictionList);
        let site = predictionList[0];
        let result = { type: 'custom', site: IMAGENET_CLASSES[site.index] }
        console.log('ML Result: Site: %s, Probability: %i%', result.site, (site.value * 100));
        if (site.value > ML_THRESHOLD) {
            return result;
        } else {
            return null;
        }
    }

    predict(image) {
            const preprocessed = this.imgToTensor(image, [224, 224])
            console.log(preprocessed);
            var result = this.model.predict(preprocessed);
            return result;
        }
        // load image to get Image object

    imgToTensor(input, size = null) {
        return tf.tidy(() => {
            let img = tf.browser.fromPixels(input);
            if (size) {
                img = tf.image.resizeBilinear(img, size);
            }
            const croppedImage = this.cropImage(img);
            const batchedImage = croppedImage.expandDims(0);
            return batchedImage.toFloat().div(tf.scalar(127)).sub(tf.scalar(1));
        });

    cropImage(img) {
        const size = Math.min(img.shape[0], img.shape[1]);
        const centerHeight = img.shape[0] / 2;
        const beginHeight = centerHeight - (size / 2);
        const centerWidth = img.shape[1] / 2;
        const beginWidth = centerWidth - (size / 2);
        return img.slice([beginHeight, beginWidth, 0], [size, size, 3]);
    };

**Comment by rthadur** [comment]
Hello, this sort of question is better suited for StackOverflow. This issues list is for bugs and feature requests. Thanks and please use the tag 'tensorflow.js' on your StackOverflow question to makes sure it's discoverable. Please use latest version of TFJS to get better results.

**Comment by tafsiri** [comment]
@plat2eau When you say the prediction accuracy is low in tfjs, are you referring to the Mobilenet model or the custom model you have on top of that? If you are getting different results for the same mobilenet model, could you make a minimal hosted example (on something like codepen or glitch) that shows different predictions between ML5 and tfjs? 

If its the latter then I agree that this is probably a question better suited for stack overflow. But I also think you might want to narrow down where the difference you are seeing emerges. You should also check that you are using the exact same inputs in both cases.

**Comment by plat2eau** [comment]
@tafsiri Actually the problem was that I was only using the custom model to make predictions and it wasn't being filtered by the base model. I corrected the code by first getting predictions from the base model and then passing the predicted tensors as an input for predicting through custom model. I'm getting pretty good results after that. 

I posted it on stack overflow and answered it [here](https://stackoverflow.com/a/57851651/11796570).
